src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -428,24 +442,52 @@ internal Task AbortAsync()
 
         private void KeepAliveTick()
         {
-            var timestamp = Stopwatch.GetTimestamp();
+            var currentTime = DateTime.UtcNow.Ticks;
+
             // Implements the keep-alive tick behavior
             // Each tick, we check if the time since the last send is larger than the keep alive duration (in ticks).
             // If it is, we send a ping frame, if not, we no-op on this tick. This means that in the worst case, the
             // true "ping rate" of the server could be (_hubOptions.KeepAliveInterval + HubEndPoint.KeepAliveTimerInterval),
             // because if the interval elapses right after the last tick of this timer, it won't be detected until the next tick.
-            if (timestamp - Interlocked.Read(ref _lastSendTimestamp) > _keepAliveDuration)
+
+            if (currentTime - Volatile.Read(ref _lastSendTimeStamp) > _keepAliveInterval)
             {
                 // Haven't sent a message for the entire keep-alive duration, so send a ping.
                 // If the transport channel is full, this will fail, but that's OK because
                 // adding a Ping message when the transport is full is unnecessary since the
                 // transport is still in the process of sending frames.
                 _ = TryWritePingAsync();
 
-                Interlocked.Exchange(ref _lastSendTimestamp, timestamp);
+                // We only update the timestamp here, because updating on each sent message is bad for performance
+                // There can be a lot of sent messages per 15 seconds
+                Volatile.Write(ref _lastSendTimeStamp, currentTime);
+            }
+        }
+
+        internal void StartClientTimeout()
+        {
+            if (_clientTimeoutActive)
+            {
+                return;
+            }
+            _clientTimeoutActive = true;
+            Features.Get<IConnectionHeartbeatFeature>()?.OnHeartbeat(state => ((HubConnectionContext)state).CheckClientTimeout(), this);
+        }
+
+        private void CheckClientTimeout()
+        {
+            // If it's been too long since we've heard from the client, then close this
+            if (DateTime.UtcNow.Ticks - Volatile.Read(ref _lastReceivedTimeStamp) > _clientTimeoutInterval)
+            {
+                Abort();
             }
         }
 
+        internal void ResetClientTimeout()
+        {
+            Volatile.Write(ref _lastReceivedTimeStamp, DateTime.UtcNow.Ticks);
```

Comment:
Last update I heard, we were going to set a bool here and do the needful in the heartbeat etc.

Reply:
Ideally this would happen. The problem is converting from a known time interval (15s) to some number of heartbeats. There's no clean way to see the heartbeat interval, and adding a field to the ```IHeartBeat``` interface would be a breaking change. Without a way to control the interval relative to what the client expects, the pattern is kinda stuck.

And so it's back to the original implementation, of just calling every 15s regardless.